cardinal/search/search.go

```diff
@@ -154,3 +226,59 @@ func (s *Search) evaluateSearch() []types.ArchetypeID {
 	cache.seen = s.reader.ArchetypeCount()
 	return cache.archetypes
 }
+
+func (s *Search) And(otherSearch *Search) *Search {
```

Comment:
These logical operators (Or/And/Not) produce invalid behavior. The boolean algebra doesn't check out if you write up the truth table.

Here's an example: https://docs.google.com/spreadsheets/d/1LHyQPTGi3mwpJYWARaU_CCtk6PczTjfro79q3hpG_zw/edit?usp=sharing

Reply:
Hey @smsunarto I took a quick look at your spread sheet. I tried to implement line 2 and it worked. See cardinal/search_test.go in this PR. 

Not sure if that's how you did the query. Let's connect Monday to see what you're talking about.

Reply:
Additionally it's a bit wierd the way the spread sheet describes what's going on. 

Take a look at H and I. H filters for component types. I filters for component properties. 

while H can exist on it's own I can only exist and modify the value of the component type filter. It does not technically make sense to say (H && I). So if an AND operator was used to connect the two broken queries there might be problems. This is untested. It's likely better to force a giant panic so that these queries can be used correctly.

Reply:
nm about the earlier comments. I see what's going on here. I'll have to rethink the operators.

Reply:
I'm thinking the component property filters just have to be all permanently anded together under every search operator.

Reply:
Operators are now only allowed to work on Searches without the where clause. 

This makes it so that the Where clause can only be called at the tail end of the search. This prevents predicates from ever being anded/ored.

Reply:
The solution for this is pretty straightforward, just and/or/not the resulting entity ids after executing the searches separately, i already wrote this down in the spreadsheet [](edited-with-Axolo)

Reply:
Moved this reply here from below: 

I must have missed that. I'm out right now so can't conveniently view your
spreadsheet. Will look when available. I thought about this route though.
The reason why I didn't go with it is because it's not clear what it means
to NOT/AND/OR the entity id or a collection of entity ids. For example NOT
calls for an inverse value but what is an inverse of an id number? A
negative number? Those were my initial thoughts.

I will take a look at what you wrote when I can.

Reply:
I took another look. The excel sheet is not entirely clear to me. 

Are you implying that the Boolean operators become set operators at the search level?

Or = union
And = intersection 
Not = complement

I'm a bit against this. It involves additional processing (looping).

Reply:
It's clear to me and I'm pretty sure anyone else on what it means. It just means that the entity is in search 1 AND/OR in search 2. The reason why your not example doesn't make sense is because it's on the same class as AND/OR operators. NOT is a negation, so you need to use it with AND/OR and not by itself.

And/or boolean algebra is analagous to union/intersect, this is just how it works in math.

I don't see any other way to do it other than this; the way you implemented produces results that doesn't make intuitive sense [](edited-with-Axolo)

Reply:
That aside, I don't think And/Or should be implemented as methods of search. 

It's impossible to get the two searches aligned using indentation, so it's not as readable. It should probably just be a top level function. [](edited-with-Axolo)

Reply:
> It's clear to me and I'm pretty sure anyone else on what it means. It just means that the entity is in search 1 AND/OR in search 2. The reason why your not example doesn't make sense is because it's on the same class as AND/OR operators. NOT is a negation, so you need to use it with AND/OR and not by itself.
> 
> And/or boolean algebra is analagous to union/intersect, this is just how it works in math.
> 
> I don't see any other way to do it other than this; the way you implemented produces results that doesn't make intuitive sense

Apologies, when I said it's not entirely clear to me I meant that in a completely neutral way. Not saying anything about the quality of your explanation or about me or about whether others understand it or not. More in the sense of "I don't get it, I'll have to talk to you about it for more clarity on my end." That's it. Honestly I wasn't even thinking about anything else, but apologies it didn't come off this way. 

The isomorphism between SET and boolean actually exists beyond just SET and Bool. It works for number theory and even graph theory. It even exists for cryptography primitives which you would likely know better than me. It's just abstract algebra and it's everywhere. Personally, I didn't fully understand what was going on with your sheets because there were expressions in that sheet where using the same symbols are used for boolean and set operators in a single expression. That's just me though, don't worry about it. 

Additionally note that NOT does work as a Set operator independently of other operators. If you have 3 entities total in the universe: [1,2,3], then NOT(1) = [2,3], there isn't a special case here. Basically if there's a NOT operator the search has to go into a secondary loop to remove relevant entities from a copy of the totality of entities to generate the complement. 

But this is besides the point. All I'm saying is that introducing a new type of operator that operates on SETs leads to a additional algorithmic complexity overhead. You are initializing a second loop for NOT and OR. Only AND does not need a secondary loop. It's also redundant on top of filters too, that is unless you want to get rid of the algebra library on filters all together? 

My personal opinion is the way I designed it now (post your post with the sheets) where the ordering of the methods makes it so operators have to be used before the WHERE clauses makes it exactly identical to how it's used before search V2 is more efficient as we won't need to launch a new loop. 

We can go whatever. I'm cool with whatever like always. This is your call, I'm just offering my perspective. 

> That aside, I don't think And/Or should be implemented as methods of search.
> 
> It's impossible to get the two searches aligned using indentation, so it's not as readable. It should probably just be a top level function.

Yeah this is just aesthetic. I agree stylistically with this, i prefer functional programming over OOP anyway.

Reply:
How about I just implement the functional operators on entity ids and I kill the filter library in subsequent PR because it becomes redundant.

Reply:
All of this is unclear to me. I need to see a properly written down spec/reference implementation if you want me to consider any alternates [](edited-with-Axolo)

Reply:
@smsunarto design doc:

https://www.notion.so/arguslabs/Alternative-Operations-for-Search-9359dd288e06471795dc1771bda30fcc?pvs=4

Reply:
Changes in the discussion in the notion doc have been implemented.

Reply:
@smsunarto I didn't realize you introduced ANOTHER proposal within my proposal. I took a quick glance and I thought it was just an example of the existing proposal. I'm getting pulled in multiple directions here. I'll take a deeper look. 

I would recommend next time if there's a new proposal, preface it with a description rather then just a code example. I need some context here on why and what are the new changes.

Reply:
Well you didn't ask any questions, you just assumed what you wanted to believe and ran with it. [](edited-with-Axolo)

Reply:
Yeah I did, but a random example in my proposal that's actually an entirely NEW design wasn't expected either [](edited-with-Axolo)

Reply:
I'm sort of getting pulled in multiple directions here [](edited-with-Axolo)

Reply:
there IS a design doc here btw and this isn't changing any of my issues [](edited-with-Axolo)

Reply:
there's two proposals, and a third proposal within my second proposal [](edited-with-Axolo)

Reply:
anyway no worries, I'll take a look and see what I can do [](edited-with-Axolo)

Reply:
just a little confused here with all the designs [](edited-with-Axolo)